### PR TITLE
optional query tests

### PR DESCRIPTION
**Pull request by jaredh159**
I mentioned in #448 that I was thinking of writing some exploratory tests to document and understand the behavior around SQL generation for nullable fields and nullable enums.

I started a branch today to start adding some simple tests, and ran into an issue I wanted some input on before I went further.  The issue is that to make assertions about the SQL strings generated by certain queries, I need to enable _deterministic hashing_  with `SWIFT_DETERMINISTIC_HASHING=1` to get reliable ordering of fields.

Unit tests are exactly the use case that this env variable seems to have been created to support, so I don't see any issue using it for the tests.  I added it to the github workflow files, as you can see in the PR. I think my hesitation stems from the fact that, if someone wants to run the tests in Xcode (I personally prefer running tests from the command line, where it's easy to set the env var) then they'll see intermittent failures with this type of test, unless they edit the scheme to add the environment variable to to the test target.  I hoped that somehow setting that variable would produce some sort check-in-able artifact in the code like I think it does in a a non-SPM Xcode project, but it doesn't seem to.

Can someone comment on whether this is acceptable? Or is there a workaround for setting the env var in code somewhere that would not require the editing of the scheme?  It seems like being able to assert on SQL generated from insert statements is pretty important.

### CODE REVIEW

**Comment by 0xTim** [comment]
Personally I don't see an issue with this. You could always wrap the tests in a test case that asserts that the environment variable is set so anyone running the tests in Xcode has a hint on what to set. 

@gwynne thoughts

**Review by gwynne — CHANGES_REQUESTED** [review]
I have no objection, actually seems like we should have already been using this, would've saved Tanner a good bit of trouble in some of his tests along the way I think 😆.

However, shell-style prefixing of commands with environment vars is not the ideal way to go in a GitHub Actions workflow (at least, not usually) - add an entry to the appropriate `env:` section instead. In this case, I'd say just add it to the top-level global environment for all CI (lines 7-8 of `test.yml`); we only really need it in those three `swift` commands, but I'll be overhauling the CI again later anyway, so seems simpler to just do it the easy way 😅

**Comment by jaredh159** [comment]
thanks @gwynne  and @0xTim.  appreciate the input.  yup, I thought about using the `env:` section instead, that seems cleaner. i'll make that change.

i'm a bit confused why these tests were passing on my machine, but failed consistently in CI... I don't suppose there is any low-level permissions you could grant me that would allow me to run CI tests in my PR without bugging you, is there?

**Comment by gwynne** [comment]
> i'm a bit confused why these tests were passing on my machine, but failed consistently in CI... I don't suppose there is any low-level permissions you could grant me that would allow me to run CI tests in my PR without bugging you, is there?

For some of the tests, it's because the CI runs against the full range of Swift versions Vapor supports - for example, the 5.2 checks are failing because `#filePath` doesn't exist until 5.3. In other cases, such as the dependents failures, the use of deterministic hashing actually changes the expected results in the `FluentBenchmarker` test cases. These cases don't run as part of FluentKit's own test suite (the ones you get when you do `swift test`); they're invoked from the various `fluent-*-driver` test suites.

And no, unfortunately there's nothing I can do to give you more access to work with the CI; GitHub Actions don't really _have_ any access beyond what you see. The most any of us can do is hit the "Re-run jobs" button.

**Comment by jaredh159** [comment]
@gwynne Thanks for the additional context regarding the test failures.

> And no, unfortunately there's nothing I can do to give you more access to work with the CI; GitHub Actions don't really _have_ any access beyond what you see. The most any of us can do is hit the "Re-run jobs" button.

About access: I didn't mean deeper access into the Github actions, I just meant that I'd like to be able to trigger the workflows to run.  When I push up a PR or a commit, the CI doesn't run, it requires approval from someone else.  I didn't realize the tests were failing until Tim (I assume) clicked to approve CI to run.

**Comment by gwynne** [comment]
> About access: I didn't mean deeper access into the Github actions, I just meant that I'd like to be able to trigger the workflows to run. When I push up a PR or a commit, the CI doesn't run, it requires approval from someone else. I didn't realize the tests were failing until Tim (I assume) clicked to approve CI to run.

Oh, that! 🙂 That's something GitHub started doing when someone opens a PR for the first time on a repository they've never contributed to before; now that Tim approved you, the CI should run automatically from here on out.

**Comment by jaredh159** [comment]
> CI should run automatically from here on out.

Just pushed a commit to test that, but it's still requiring approval.  I'm guessing I need to actually land a commit in this repo for that to not be the case.  Not a huge deal.  I'll maybe temporarily make an unforked copy to run the GH actions myself until i get it sorted out and get a commit merged.

**Comment by gwynne** [comment]
Hm. Oh well - I'll keep an eye on it for now 😅

**Comment by 0xTim** [comment]
Yeah until you get your first contributions merged in it needs to be approved each time unfortunately

File: Tests/FluentKitTests/OptionalEnumQueryTests.swift
Comment (jaredh159) [reply Tests/FluentKitTests/OptionalEnumQueryTests.swift]: Yeah, it turns out it's not supported on Linux, and not before Xcode 12.5, so I temporarily skipped these tests by prefixing the test method.  Or, I could remove them if you prefer.  But I'll have at least one PR coming after this is merged (hopefully) fixing one of these, and maybe another shortly after.

Comment (jaredh159) [reply Tests/FluentKitTests/OptionalEnumQueryTests.swift]: @calebkleveter updated 👍

**Comment by jaredh159** [comment]
@gwynne moved `SWIFT_DETERMINISTIC_HASHING=1` into `env:` section, as requested, sorry forgot about that earlier.

**Review by gwynne — APPROVED** [review]
LGTM!

**Comment by 0xTim** [comment]
@gwynne anything blocking this from being merged?

**Pull request merged**
